src/data_type/mod.rs

```diff
@@ -2709,11 +2725,17 @@ where
 impl Or<DataType> for DataType {
     type Sum = DataType;
     fn or(self, other: DataType) -> Self::Sum {
-        // Simplify in the case of struct and Unit
-        match self {
-            DataType::Null => other,
-            DataType::Union(u) => u.or(other).into(),
-            s => Union::from_data_type(s).or(other).into(),
+        // Simplify in the case of struct
```

Comment:
For consistency only the match on self is done here, the match on the second arg is done in the implementation of the variant.

Reply:
This means `Unit`, `Option` and `Union` should implement `Or` and you should only match `self` against `Unit`, `Option` or `Union` (or non tuple variant such as `Null` or `Any` if relevant) and delegate to the Variant implementation.